### PR TITLE
User-facing dictionaries as Pydantic models

### PR DESCRIPTION
**Pull request by jlnav**
Exploration of defining many user-facing data structures as [Pydantic](https://pydantic-docs.helpmanual.io/) models, which can perform type and valid-field checking _on instantiation_. This approach should be more robust than performing dozens of ``assert isinstance(...`` throughout our code.

Users will be able to use either these structures _or_ the traditional dictionary approach. I'm thinking that behind the scenes I'll try casting input dictionaries to valid models to check validity.

Don't panic, we can still add necessary validators like checking if out-fields from a GenSpec are included in the in-fields of a SimSpec.

Should make redundant much of the logic within `check_inputs.py`

Addresses #878

### CODE REVIEW

**Comment by jlnav** [comment]
Example docs: https://libensemble--874.org.readthedocs.build/en/874/data_structures/sim_specs.html

File: docs/data_structures/gen_specs.rst [hunk 1]
```diff
@@ -1,59 +1,80 @@
...
-            field names (as strings) that will be given back to a persistent gen_f
-        'out' [list of tuples (field name, data type, [size])]:
-            gen_f outputs that will be stored in the libEnsemble history
-        'user' [dict]:
-            Data structure to contain problem specific constants and/or input data
+Generator Specs
+===============
+
+Used to specify the generator function, its inputs and outputs, and user data. Can be constructed and passed to libEnsemble
+as a Python class (with automatic input validation) or a simple dictionary.
```
Comment (shuds13) [inline docs/data_structures/gen_specs.rst:7]: Is input validation used with the dictionary also? This wording to me suggests only when provided as a class.

File: libensemble/libE.py [hunk 2]
```diff
@@ -175,20 +178,28 @@
...
+    alloc_specs: AllocSpecs = alloc_defaults.alloc_specs,
+    libE_specs: LibeSpecs = {},
+    H0=None,
+) -> (np.ndarray, Dict, int):
     """
     Parameters
     ----------
 
-    sim_specs: :obj:`dict`
+    sim_specs: :obj:`dict` or :obj:`SimSpecs`
```
Comment (jmlarson1) [inline libensemble/libE.py:202]: Should these things be one " ` " or two " `` " ?

**Comment by jlnav** [comment]
More touched files today due to:

- Converting most single-quotes to double-quotes in code-samples throughout docs
- Moving History array doc to toctree directly underneath Constructing Workflows
- Converting many :doc: tags to :ref: tags so links work even when files moved, and subsections can be linked to
- Various other clarifications

**Review by jmlarson1 — APPROVED** [review]

File: docs/advanced_installation.rst [hunk 3]
```diff
@@ -8,7 +8,7 @@ offer a trade-off between convenience and the ability
 to customize builds, including platform-specific optimizations.
 
 Further recommendations for selected HPC systems are given in the
-:doc:`HPC platform guides<platforms/platforms_index>`.
+:ref:`HPC platform guides<platform-index>`.
```
Comment (shuds13) [inline docs/advanced_installation.rst:11]: Dont see link in docs

File: docs/data_structures/sim_specs.rst [hunk 4]
```diff
@@ -1,44 +1,66 @@
...
+    sim_specs = {
+      "sim_f": six_hump_camel,
+      "in": ["x"],
+      "out": [("y", float)],
+    }
+    ...
+
+.. _sim-specs-example1:
+
+  - test_uniform_sampling.py_ has a :class:`sim_specs<libensemble.specs.SimSpecs>`  that declares
```
Comment (shuds13) [inline docs/data_structures/sim_specs.rst:53]: broken link - this should really be a regression test?

File: docs/running_libE.rst [hunk 5]
```diff
@@ -89,28 +91,13 @@ TCP Comms
 
 The TCP option can be used to run the Manager on one system and launch workers to remote
 systems or nodes over TCP. The necessary configuration options can be provided through
-``libE_specs``, or on the command line if you are using the :doc:`parse_args()<utilities>` function.
```
Comment (shuds13) [inline docs/running_libE.rst:92]: Is this removed or moved?

File: docs/data_structures/sim_specs.rst (hunk 4 above)
Comment (jmlarson1) [reply docs/data_structures/sim_specs.rst:53]: @shuds13 Is there an automated way to test that links aren't broken?

File: docs/data_structures/data_structures.rst [hunk 6]
```diff
@@ -1,45 +1,36 @@
...
-:ref:`generation<datastruct-gen-specs>`,
-and
-:ref:`allocation<datastruct-alloc-specs>`
-specifications, as well as specify
-:ref:`exit criteria<datastruct-exit-criteria>`,
-:ref:`persistent information<datastruct-persis-info>`, and other
-:ref:`libEnsemble<datastruct-libe-specs>`
-options.
+This section outlines the dictionaries, classes, and other structures used by libEnsemble
+to configure a workflow, oftentimes within a single Python file we refer to as a *calling script*.
```
Comment (shuds13) [inline docs/data_structures/data_structures.rst:5]: Maybe could mention the YAML interface here as an alternative?

**Review by shuds13 — APPROVED** [review]

**Pull request merged**